gadget/device_test.go

```diff
@@ -0,0 +1,205 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package gadget_test
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	. "gopkg.in/check.v1"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/gadget"
+)
+
+type deviceSuite struct {
+	dir string
+}
+
+var _ = Suite(&deviceSuite{})
+
+func (d *deviceSuite) SetUpTest(c *C) {
+	d.dir = c.MkDir()
+	dirs.SetRootDir(d.dir)
+
+	err := os.MkdirAll(filepath.Join(d.dir, "/dev/disk/by-label"), 0755)
+	c.Assert(err, IsNil)
+	err = os.MkdirAll(filepath.Join(d.dir, "/dev/disk/by-partlabel"), 0755)
+	c.Assert(err, IsNil)
+	err = ioutil.WriteFile(filepath.Join(d.dir, "/dev/fakedevice"), []byte(""), 0644)
+	c.Assert(err, IsNil)
+}
+
+func (d *deviceSuite) TearDownTest(c *C) {
+	dirs.SetRootDir("/")
+}
+
+func (d *deviceSuite) TestDeviceFindByStructureName(c *C) {
+	names := []struct {
+		escaped   string
+		structure string
+	}{
+		{"foo", "foo"},
+		{"123", "123"},
+		{"foo\\x20bar", "foo bar"},
+		{"foo#bar", "foo#bar"},
```

Comment:
Nitpick, I think this test (and the one below) could also check a multi-byte utf label case, for completeness.